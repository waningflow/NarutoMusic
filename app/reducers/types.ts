// 定义：type, payload, action, dispatch
import { Dispatch, Store } from 'redux';

export type User = {
  status: UserStatus;
};

export enum UserStatus {
  UNLOG = 'not-logged',
  LOGGING = 'logging',
  LOGGED = 'logged-in'
}

export const UPDATE_USER = 'UPDATE_USER';

interface UpdateUserAction {
  type: typeof UPDATE_USER;
  payload: User;
}

export type UserAction = UpdateUserAction;

export type UserDispatch = Dispatch<UserAction>;

export type UserStore = Store<User, UserAction>;
// export type counterStateType = {
//   counter: number;
// };

// export type GetState = () => counterStateType;

// export type Dispatch = ReduxDispatch<Action<string>>;

// export type Store = ReduxStore<counterStateType, Action<string>>;
