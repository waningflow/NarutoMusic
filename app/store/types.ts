// 定义：type, payload, action, dispatch
import { Dispatch } from 'redux';
import { User, Playlist } from '@/types';

// user
export const UPDATE_USER = 'UPDATE_USER';

export interface UpdateUserAction {
  type: typeof UPDATE_USER;
  payload: Partial<User>;
}

export type UserAction = UpdateUserAction;

export type UserDispatch = Dispatch<UserAction>;

// playlist
export const UPDATE_PLAYLIST = 'UPDATE_PLAYLIST';

export interface UpdatePlaylistAction {
  type: typeof UPDATE_PLAYLIST;
  payload: Playlist;
}

export type PlaylistAction = UpdatePlaylistAction;

export type PlaylistDispatch = Dispatch<UpdatePlaylistAction>;
