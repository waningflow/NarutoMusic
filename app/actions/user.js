export const LOGIN_STATUS_UPDATE = 'LOGIN_STATUS_UPDATE';

export function updateStatus(status: string) {
  return {
    type: LOGIN_STATUS_UPDATE,
    value: status
  };
}

export function login() {
  return dispatch => {
    dispatch({
      type: LOGIN_STATUS_UPDATE,
      value: 'loging'
    });
    setTimeout(() => {
      dispatch({
        type: LOGIN_STATUS_UPDATE,
        value: 'loged'
      });
    }, 1000);
  };
}
