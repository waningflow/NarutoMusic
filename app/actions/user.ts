import {
  UPDATE_USER,
  UserStatus,
  UserAction,
  UserDispatch
} from '../reducers/types';

import { loginWithCellphone } from '../api';

export function updateUserStatus(status: UserStatus): UserAction {
  return {
    type: UPDATE_USER,
    payload: {
      status
    }
  };
}

export function login(params: any) {
  return async (dispatch: UserDispatch) => {
    dispatch(updateUserStatus(UserStatus.LOGGING));
    // await sleep(1000);
    try {
      await loginWithCellphone(params);
      dispatch(updateUserStatus(UserStatus.LOGGED));
    } catch (e) {
      dispatch(updateUserStatus(UserStatus.UNLOG));
    }
  };
}
