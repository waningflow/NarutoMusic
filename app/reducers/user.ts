import { UserStatus, UPDATE_USER, UserAction } from './types';

const initState = {
  status: UserStatus.UNLOG
};

export default function user(state = initState, action: UserAction) {
  switch (action.type) {
    case UPDATE_USER:
      return { ...state, ...action.payload };
    default:
      return state;
  }
}
