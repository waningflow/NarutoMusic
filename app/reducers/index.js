// @flow
import { combineReducers } from 'redux';
import { connectRouter } from 'connected-react-router';
import type { HashHistory } from 'history';
import user from './user';

export default function createRootReducer(history: HashHistory) {
  return combineReducers<{}, *>({
    router: connectRouter(history),
    user
  });
}
