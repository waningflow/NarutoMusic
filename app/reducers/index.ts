import { combineReducers, Store as ReduxStore, Action } from 'redux';
import router from './history';
import user from './user';
import playlist from './playlist';

const rootReducer = combineReducers({
  router,
  user,
  playlist
});
export default rootReducer;

export type State = ReturnType<typeof rootReducer>;

export type Store = ReduxStore<State, Action<string>>;
