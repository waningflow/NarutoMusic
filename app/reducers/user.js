import { LOGIN_STATUS_UPDATE } from '../actions/user';

const initState = {
  status: 'unloged'
};

export default function counter(state = initState, action) {
  switch (action.type) {
    case LOGIN_STATUS_UPDATE:
      return Object.assign({}, state, { status: action.value });
    default:
      return state;
  }
}
