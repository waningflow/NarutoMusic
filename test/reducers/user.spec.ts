import user from '../../app/reducers/user';
import { UPDATE_USER, UserStatus } from '../../app/reducers/types';

describe('reducers', () => {
  describe('user', () => {
    it('should handle initial state', () => {
      expect(user(undefined, {})).toMatchSnapshot();
    });

    it('should handle LOGIN_STATUS_UPDATE', () => {
      expect(
        user(undefined, {
          type: UPDATE_USER,
          payload: { status: UserStatus.LOGGING }
        })
      ).toMatchSnapshot();
    });
  });
});
