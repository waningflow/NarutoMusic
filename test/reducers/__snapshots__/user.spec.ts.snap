// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers user should handle LOGIN_STATUS_UPDATE 1`] = `
Object {
  "status": "logging",
}
`;

exports[`reducers user should handle initial state 1`] = `
Object {
  "status": "not-logged",
}
`;
